From 16fe942a4226204ab613d9b477c8db93b3cee0b4 Mon Sep 17 00:00:00 2001
From: Evgeny Vereshchagin <evvers@ya.ru>
Date: Fri, 19 Jun 2026 02:52:13 +0000
Subject: [PATCH] core: add missing check to avahi_record_copy

to address
```
fuzz-consume-record-nallocfuzz: fuzz/fuzz-consume-record.c:95: int LLVMFuzzerTestOneInput(const uint8_t *, size_t): Assertion `ret' failed.
==655035== ERROR: libFuzzer: deadly signal
    #0 0x000000551fd5 in __sanitizer_print_stack_trace (/home/vagrant/avahi/out/fuzz-consume-record-nallocfuzz+0x551fd5) (BuildId: abb5fbc4de2d9ff171faf1dbf07
1d1462ca11667)
    #1 0x00000043b1ec in fuzzer::PrintStackTrace() fuzzer.o
    #2 0x00000041d897 in fuzzer::Fuzzer::CrashCallback() fuzzer.o
    #3 0x7f1a202dd0af  (/lib64/libc.so.6+0x1a0af) (BuildId: 18472003bbf1c5f098a09b5016b9b8bd4c7c59f0)
    #4 0x7f1a20337dcb in __pthread_kill_implementation (/lib64/libc.so.6+0x74dcb) (BuildId: 18472003bbf1c5f098a09b5016b9b8bd4c7c59f0)
    #5 0x7f1a202dcf8d in gsignal (/lib64/libc.so.6+0x19f8d) (BuildId: 18472003bbf1c5f098a09b5016b9b8bd4c7c59f0)
    #6 0x7f1a202c47b2 in abort (/lib64/libc.so.6+0x17b2) (BuildId: 18472003bbf1c5f098a09b5016b9b8bd4c7c59f0)
    #7 0x7f1a202c5803 in __libc_message_impl.cold (/lib64/libc.so.6+0x2803) (BuildId: 18472003bbf1c5f098a09b5016b9b8bd4c7c59f0)
    #8 0x7f1a202d4e44 in __assert_fail (/lib64/libc.so.6+0x11e44) (BuildId: 18472003bbf1c5f098a09b5016b9b8bd4c7c59f0)
    #9 0x00000058da72 in LLVMFuzzerTestOneInput /home/vagrant/avahi/fuzz/fuzz-consume-record.c:95:5
```
---
 avahi-core/rr.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/avahi-core/rr.c b/avahi-core/rr.c
index 73c1b84..5c40c9f 100644
--- a/avahi-core/rr.c
+++ b/avahi-core/rr.c
@@ -456,7 +456,8 @@ AvahiRecord *avahi_record_copy(AvahiRecord *r) {
             break;
 
         case AVAHI_DNS_TYPE_TXT:
-            copy->data.txt.string_list = avahi_string_list_copy(r->data.txt.string_list);
+            if (r->data.txt.string_list && !(copy->data.txt.string_list = avahi_string_list_copy(r->data.txt.string_list)))
+                goto fail;
             break;
 
         case AVAHI_DNS_TYPE_A:
